### PR TITLE
Add first draft of ZXTransformer to contrib.

### PR DESCRIPTION
**Pull request by dlyongemallo**
This is a custom transformer which uses ZX-calculus through the PyZX library to perform circuit optimisation.

See issue #6585.

### CODE REVIEW

File: cirq-core/cirq/contrib/zxtransformer/zxtransformer.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+    def __init__(self, optimize: Optional[Callable[[zx.Circuit], zx.Circuit]] = None):
+        """Initializes transformer.
+
+        Args:
+            optimize: The optimization routine to execute. Defaults to `pyzx.simplify.full_reduce` if not specified.
+        """
+
+        super().__init__()
+        self.qubits: List[cirq.Qid] = []
```
Comment (NoureldinYosri) [inline cirq-core/cirq/contrib/zxtransformer/zxtransformer.py:93]: I don't think the qubits and qubit_to_index map should be properities of the class. they are properties of the circuit being transformed so they shouldn't be stored this way. 

Actually I don't think this transfomer need to be class. I think it should be a function
```py3
@transformer
def zx_transfomer(circuit: circuits.AbstractCircuit, context: Optional[cirq.TransformerContext] = None, optimizer: Callable[[zx.Circuit], zx.Circuit] = _optimize) -> circuits.AbstractCircuit
```
The qubits and qubit_to_index lists should be passed as arguments to the methods that need them.

**Review by NoureldinYosri — CHANGES_REQUESTED** [review]

File: cirq-core/cirq/contrib/zxtransformer/zxtransformer.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+from pyzx.circuit import gates as zx_gates
+
+
+def cirq_gate_to_zx_gate(cirq_gate: Optional[cirq.Gate], qubits: List[int]) -> Optional[zx_gates.Gate]:
+    """Convert a Cirq gate to a PyZX gate."""
+    if cirq.Gate is None:
+        return None
+
+    if isinstance(cirq_gate, (cirq.Rx, cirq.XPowGate)):
+        return zx_gates.XPhase(*qubits, phase=cirq_gate.exponent)  # type: ignore
```
[the remark this answers is not in the document]
Comment (dlyongemallo) [reply cirq-core/cirq/contrib/zxtransformer/zxtransformer.py:32]: `Expected type 'Fraction | int | Poly', got 'Float | Expr' instead.`

[the remark this answers is not in the document]
Comment (NoureldinYosri) [reply cirq-core/cirq/contrib/zxtransformer/zxtransformer.py:32]: I think this should fix it, cirq gates support symbolic properties but pyzx doesn't.

```suggestion
        return zx_gates.XPhase(*qubits, phase=float(cirq_gate.exponent)) 
```

alternatively you can inspect the type of the exponent and raise an error if it's a sympy expr

**Comment by NoureldinYosri** [comment]
@mhucka the contributer has indicated in the issue that they don't have time to work on this https://github.com/quantumlib/Cirq/issues/6585#issuecomment-2382014994

**Comment by mhucka** [comment]
That's correct, but they asked if someone else could take it over; they didn't say to close it. Moreover, the associated issue is still open. Since this PR was auto-closed by bot due to staleness, it seems like the right action is to reopen it and at least have a discussion about whether both the issue and the PR should be closed deliberately.

**Comment by dstrain115** [comment]
@naerabati Are you still interested in doing work on the ZX calculus transformer?  Want to pick this up?

**Comment by naerabati** [comment]
Sorry I couldn't answer earlier! I don't think right now I will have the time to finish the transformer either

**Comment by naerabati** [comment]
@dstrain115 @dlyongemallo would you be able to tell me what else needs to implemented to complete this PR?

**Comment by mhucka** [comment]
Closing due to inactivity.

**Comment by dlyongemallo** [comment]
I can actually pick this back up. I recently updated the change to use `pyzx` 0.10.0, which adds support for `reset` and `measurement` operations which were previously not available when this PR was first created.

What needs to be done from my end to get this merged?

**Comment by mhucka** [comment]
@dlyongemallo Thank you for your work. It looks like GitHub will not let me reopen this PR; I get an error message that the branch was "either force-pushed or recreated".

It's possible GitHub will let you reopen it. If the button to reopen the PR is available to you, go ahead and click it.

If it is not available, then the easiest thing to do would be to open a fresh PR. In the PR description, you can write something to the effect of "continuation of PR #6647" or something like that, to help people follow.

**Comment by dlyongemallo** [comment]
@mhucka I wasn't able to re-open this. I've opened a fresh PR #8026.

**Pull request closed without merging**